components/compass-runtime-agent/internal/compass/director/proxy_test.go

```diff
@@ -0,0 +1,158 @@
+package director
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSingleJoiningSlash(t *testing.T) {
+	testCases := map[string]struct {
+		proxyPath      string
+		requestPath    string
+		expectedResult string
+	}{
+		"Request path is added to proxy with adding additional slash at the end": {
+			proxyPath:      "director/graphql",
+			requestPath:    "foo/bar",
+			expectedResult: "director/graphql/foo/bar",
+		},
+		"Request path is added to proxy and keeping adding additional slash at the end": {
+			proxyPath:      "director/graphql",
+			requestPath:    "foo/bar/",
+			expectedResult: "director/graphql/foo/bar/",
+		},
+		"Trim trailing slash if request path is empty with slash": {
+			proxyPath:      "director/graphql",
+			requestPath:    "/",
+			expectedResult: "director/graphql",
+		},
+		"Do not add trailing slash if request path is empty without slash": {
+			proxyPath:      "director/graphql",
+			requestPath:    "",
+			expectedResult: "director/graphql",
+		},
+	}
+
+	for tN, tC := range testCases {
+		t.Run(tN, func(t *testing.T) {
+			result := singleJoiningSlash(tC.proxyPath, tC.requestPath)
+			assert.Equal(t, tC.expectedResult, result)
+		})
+	}
+}
+
+func TestProxyServeHTTP(t *testing.T) {
+	t.Run("return service unavailable error when proxy to director is not yet configured", func(t *testing.T) {
+		// given
+		var (
+			expectedBody = "Proxy to Director is not initialized. Try again later.\n"
+			expectedCode = http.StatusServiceUnavailable
+
+			fixRequest  = httptest.NewRequest(http.MethodPost, "http://hakuna.com/matata", nil)
+			spyResponse = httptest.NewRecorder()
+		)
+
+		proxy := NewProxy(ProxyConfig{})
+
+		// when
+		proxy.ServeHTTP(spyResponse, fixRequest)
+
+		// then
+		assert.Equal(t, expectedCode, spyResponse.Code)
+		assert.Equal(t, expectedBody, spyResponse.Body.String())
+	})
+
+	t.Run("proxying to set target URL with the given certs", func(t *testing.T) {
+		// given
+		var (
+			expectedBody = "Buenos Aires from Director!"
+			expectedCode = http.StatusOK
+			spyResponse  = httptest.NewRecorder()
+			fixCert      = MustLoadFixCert(t)
+		)
+
+		var gotPeerCertificates []*x509.Certificate
+		ts := NewTestTLSServer(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I think it would be good to close the server after the test, as mentioned in the docs:
> // The caller should call Close when finished, to shut it down.

Reply:
added